### PR TITLE
Add support for mutiny types in reactive routes

### PR DESCRIPTION
**Pull request by cescoffier**
Add support for Uni, Multi, and SSE to reactive routes.

### CODE REVIEW

File: docs/src/main/asciidoc/reactive-routes.adoc [hunk 1]
```diff
@@ -163,6 +163,189 @@ public class SimpleRoutes {
...
+@Route(path = "/hello")
+Multi<String> hellos(RoutingContext context) {
+    return Multi.createFrom().items("hello", "world", "!");  // <1>
+}
+----
+1. Produces `["hello", "world", "!"]`
+
+The method can return:
+
+* a `Multi<String>` - the items are written one by one in a JSON Array.
```
Comment (mkouba) [inline docs/src/main/asciidoc/reactive-routes.adoc:243]: Hm, does it make sense to always produce a JSON array? It's not very consistent with `Uni<String>` and `String` return types.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 2]
```diff
@@ -496,25 +473,74 @@ void implementInvoke(BeanInfo bean, MethodInfo method, ClassCreator invokerCreat
 
         ResultHandle paramHandle;
         MethodDescriptor methodDescriptor;
+        String returnType = descriptor.getReturnType().name().toString();
+
+        // TODO Make Routing Context optional, allow injected Response and Request individually.
```
Comment (mkouba) [inline extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:478]: +1 if a method returns something we should make the context param optional...

File: docs/src/main/asciidoc/reactive-routes.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: It the behavior used in other frameworks. 
For raw output, use buffer.

Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: Well, I'm talking about use cases like `io.quarkus.qute.TemplateInstance.publisher()`. 

WRT consistency - we don't encode the `String` and `Uni<String>` return values to JSON either...

Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: > It the behavior used in other frameworks.

What frameworks BWT? ;-)

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 2 above)
Comment (mkouba) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java:478]: I've created https://github.com/quarkusio/quarkus/issues/10397

**Review by mkouba — CHANGES_REQUESTED** [review]
LGTM. 

Except for the inconsistency when returning `Multi<String>` as described here: https://github.com/quarkusio/quarkus/pull/10379#discussion_r448275405. I don't see a good reason for this behavior. 

However, if there are any strong arguments feel free to dismiss my review...

File: docs/src/main/asciidoc/reactive-routes.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: Spring Routes do this, I also remember that Micronaut does something similar. 

The ideal would be to check the "produces" attribute from the Route annotation. However, at the moment, we generate one method per "method" which can have multiple Route annotations.

Comment (mkouba) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: > Spring Routes do this...

Yes, it seems that they do this based on the media type.

> The ideal would be to check the "produces" attribute from the Route annotation. However, at the moment, we generate one method per "method" which can have multiple Route annotations.

That would make perfect sense. But until we implement this I don't think it makes any sense to use some default (JSON) for `Multi<String>` but no default for `Uni<String>` and `String`.

Comment (cescoffier) [reply docs/src/main/asciidoc/reactive-routes.adoc:243]: @mkouba and I discussed it and we are going to use the same approach we use for SSE. I'm going to update the PR soon.

**Review by mkouba — APPROVED** [review]

**Comment by mkouba** [comment]
JDK 14 failure is unrelated. The `ConcurrentExecutionTest.testNonconcurrentExecution()` is unreliable. I'll need to address it in a separate PR.

**Pull request merged**